### PR TITLE
[GSK-2346] Avoid copying whole dataset when doing slicing

### PR DESCRIPTION
**Pull request by Hartorn**
## Description
DataProcessor was doing a copy every time, event in the case of slice.
The issue is the slice was actually referencing the whole dataset, so the whole dataset was kept in memory.
Now, we only reference the existing one.

Testing on another large dataset (76020, 371), was crashing before, working now.

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [ ] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CODE_OF_CONDUCT.md) document.
- [ ] I've read the [`CONTRIBUTING.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CONTRIBUTING.md) guide.
- [ ] I've updated the code style using `make codestyle`.
- [ ] I've written tests for all new methods and classes that I created.
- [ ] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

**Review by kevinmessiaen — CHANGES_REQUESTED** [review]
Fixes the issue, However we can get ride of the `copy` flag altogether since doing a copy change nothing in the behaviour of the code expect loading duplicated data in the ram

**Comment by mattbit** [comment]
@Hartorn I would be careful because I remember having to introduce this copy operation because of some bug https://github.com/Giskard-AI/giskard/commit/b092277273305acc37b50a365b17ff03a841d849.
I can't find the PR associated to that commit right now but I would check

File: giskard/datasets/base/__init__.py [hunk 1]
```diff
@@ -77,8 +79,11 @@ def add_step(self, processor: Union[SlicingFunction, TransformationFunction]):
             self.pipeline.append(processor)
         return self
 
-    def apply(self, dataset: "Dataset", apply_only_last=False, get_mask: bool = False):
-        ds = dataset.copy()
+    def apply(self, dataset: "Dataset", apply_only_last=False, get_mask: bool = False, copy: bool = True):
+        if copy:
+            ds = dataset.copy()
+        else:
+            ds = dataset
```
[the remark this answers is not in the document]
Comment (Hartorn) [reply giskard/datasets/base/__init__.py:86]: Not true, making a copy may still be necessary.
For example, in the TextTransformation, code is directly modifying the dataframe, and so copy is still needed.

To remove this copy, we need to go through every slicing function and transformation, and I don't want this pr to take too long.

**Comment by Hartorn** [comment]
> @Hartorn I would be careful because I remember having to introduce this copy operation because of some bug [b092277](https://github.com/Giskard-AI/giskard/commit/b092277273305acc37b50a365b17ff03a841d849). I can't find the PR associated to that commit right now but I would check

atm TextTransformation modify dataframe directly, so I don't want to change this beahaviour either (except when slicing)

**Review by mattbit — APPROVED** [review]
At some point we'll need to improve this, but the current fix is ok for me.

File: giskard/datasets/base/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevinmessiaen) [reply giskard/datasets/base/__init__.py:86]: Ah my bad, you're right, I overlooked this part :/

**Review by kevinmessiaen — APPROVED** [review]

**Pull request merged**